SDWebImage/SDWebImageDownloader.m

```diff
@@ -251,33 +250,24 @@ - (nullable SDWebImageDownloadToken *)addProgressCallback:(SDWebImageDownloaderP
         }
         return nil;
     }
-
-    __block SDWebImageDownloadToken *token = nil;
-
-    dispatch_barrier_sync(self.barrierQueue, ^{
-        SDWebImageDownloaderOperation *operation = self.URLOperations[url];
-        if (!operation) {
-            operation = createCallback();
-            self.URLOperations[url] = operation;
-
-            __weak SDWebImageDownloaderOperation *woperation = operation;
-            operation.completionBlock = ^{
-				dispatch_barrier_sync(self.barrierQueue, ^{
-					SDWebImageDownloaderOperation *soperation = woperation;
-					if (!soperation) return;
-					if (self.URLOperations[url] == soperation) {
-						[self.URLOperations removeObjectForKey:url];
-					};
-				});
-            };
-        }
-        id downloadOperationCancelToken = [operation addHandlersForProgress:progressBlock completed:completedBlock];
-
-        token = [SDWebImageDownloadToken new];
-        token.downloadOperation = operation;
-        token.url = url;
-        token.downloadOperationCancelToken = downloadOperationCancelToken;
-    });
+    
+    SDWebImageDownloaderOperation *operation = [self operationForURL:url];
+    if (!operation) {
+        operation = createCallback();
+        [self setOperation:operation forURL:url];
```

Comment:
I think maybe this is not thread-safe. Even this will not cause crash. But if user call `addProgressCallback` from the different queue with the same url. If thread A and thread B enter this if condition at the same time(Can happen). This will cause we create two operations but only the last one was alive(The first one will be replaced).

I think we shuold better consider use a lock again with `dispatch_semaphore` again.